planner/core/task.go

```diff
@@ -1633,11 +1641,89 @@ func (p *basePhysicalAgg) newPartialAggregate(copTaskType kv.StoreType, isMPPTas
 		MppRunMode:   p.MppRunMode,
 	}.initForHash(p.ctx, p.stats, p.blockOffset, prop)
 	finalAgg.schema = finalPref.Schema
+	// partialAgg and finalAgg use the same ref of stats
 	return partialAgg, finalAgg
 }
 
+func (p *basePhysicalAgg) scale3StageForDistinctAgg() (bool, expression.GroupingSets) {
+	if p.canUse3Stage4SingleDistinctAgg() {
```

Comment:
Maybe this logic is more clearly ?
```
if (!Enable3StageDistinctAgg) {
    return false
}
distinct_num = 0;
if (distinct_num == 1) {
    canUse3Stage4SingleDistinctAgg();
} else {
    canUse3Stage4MultiDistinctAgg();
}
```
(BTW, this is just my personal opinion)

Reply:
yes, switch judgment can be taken out here, while the distinct_num can only be scratched inside